src/mame/rm/rm380z_v.cpp

```diff
@@ -141,33 +138,109 @@ void rm380z_state::decode_videoram_char(int row, int col, uint8_t& chr, uint8_t
 	}
 }
 
-void rm380z_state::scroll_videoram()
+void rm380z_state::scroll_videoram_up(int n, bool clear)
 {
-	// scroll up one row of videoram
-	std::memmove(m_vramchars, m_vramchars[1], (RM380Z_SCREENROWS - 1) * RM380Z_SCREENCOLS);
-	std::memmove(m_vramattribs, m_vramattribs[1], (RM380Z_SCREENROWS - 1) * RM380Z_SCREENCOLS);
+	// move last n rows of videoram up one
+	std::memmove(m_vramchars[RM380Z_SCREENROWS - (n + 1)], m_vramchars[RM380Z_SCREENROWS - n], n * RM380Z_SCREENCOLS);
+	std::memmove(m_vramattribs[RM380Z_SCREENROWS - (n + 1)], m_vramattribs[RM380Z_SCREENROWS - n], n * RM380Z_SCREENCOLS);
 
-	// the last line is filled with spaces
-	std::memset(m_vramchars[RM380Z_SCREENROWS - 1], 0x20, RM380Z_SCREENCOLS);
-	std::memset(m_vramattribs[RM380Z_SCREENROWS - 1], 0x00, RM380Z_SCREENCOLS);
+	if (clear)
+	{
+		// the last line is filled with spaces
+		std::memset(m_vramchars[RM380Z_ROW_MAX], 0x20, RM380Z_SCREENCOLS);
+		std::memset(m_vramattribs[RM380Z_ROW_MAX], 0x00, RM380Z_SCREENCOLS);
+	}
 }
 
+void rm380z_state::scroll_videoram_down(int n, bool clear)
+{
+	// move first n rows of videoram down one
+	std::memmove(m_vramchars[RM380Z_SCREENROWS - n], m_vramchars[RM380Z_SCREENROWS - (n + 1)], n * RM380Z_SCREENCOLS);
+	std::memmove(m_vramattribs[RM380Z_SCREENROWS - n], m_vramattribs[RM380Z_SCREENROWS - (n + 1)], n * RM380Z_SCREENCOLS);
+
+	if (clear)
+	{
+		// the first line is filled with spaces
+		std::memset(m_vramchars[RM380Z_SCREENROWS - (n + 1)], 0x20, RM380Z_SCREENCOLS);
+		std::memset(m_vramattribs[RM380Z_SCREENROWS - (n + 1)], 0x00, RM380Z_SCREENCOLS);
+	}
+}
```

Comment:
In reality, the hardware won’t copy the contents of video memory like this – it will use the contents of the scroll registers to adjust addressing.  Rather than further bake in the flawed approach, can you please rework the code to use the scroll registers to adjust addressing when accessing video memory and when drawing?

You can see examples of this in source/mame/osborne/osborne1.cpp and src/mame/tandberg/tdv2100_disp_logic.cpp (although they only support full screen scrolling).

Reply:
I can certainly take a look at that, but will have to do some research.  Would it then in effect mean that after scrolling future screen writes to row 23 would be stored m_vramchars[0] and the scroll registers would tell us that row 0 maps to m_vramchars[1] and row 23 maps to m_vramchars[0] etc?

By the way I've just discovered another use for FBFD.  It's used to send character data when changing the user definable characters when bit 3 of FBFC is set - another unimplemented feature that I am looking into!

Reply:
I've now got a scroll register working for full screen and it does seem a lot better.  I will continue working on this.

Reply:
Please review latest commit.  I was able to use a register for full screen scrolling and that was a great idea.  However, the 4-line scrolling is odd because the firmware (COS) copies one of the lines itself.  After telling the HW to scroll it copies line 23 to line 22, and then clears line 23.  For this to work the HW must be changing lines 20 and 21?  This is what I am now doing, but it's annoying.  I did have a second scroll register setup, but alas I don't think it's possible to use that.  What do you think?

Reply:
Actually I think it is possible with 3 registers, but I'll need to spend a little more time on the algorithm...

Reply:
After banging my head against a wall thinking it was impossible I finally realised that 4-line HW scrolling is impossible and it's actually mostly done by the firmware!  However, to help it do that it uses full screen scrolling to shift things around.  This is why we see the scroll counter increment and then decrement.  I put some trace to show this alongside vram access and it suddenly made sense :-)  This is what we now see when moving up to scroll text down (lines held "1", "2", "3", and "4"):

```
Reading row: 22, Data: 3
Writing row: 23, Data: 3
Scrolling down
Reading row: 22, Data: 2
Writing row: 23, Data: 2
Scrolling up
Reading row: 20, Data: 1
Writing row: 21, Data: 1
Writing row: 20, Data: (blank char)
```

So with just one scroll register we can do everything and it's really simple.  No need to know or care if the firmware is trying to do a 4-line scroll or not.  As long as we implement full screen scrolling everything else just works.  Thank you for the great suggestion.  I've certainly learned something anyway.

Please check the latest version.  Btw - some of the automated checks seem to fail but they are not related to my change.  I think the build must have been a bit broken when I made the branch.

Reply:
One further revelation is that the value we get from port FBFD is probably a scroll register value maintained by the firmware.  I think we can use that directly, e.g.

```
	case 0xfd:      // screen line counter (?)
		//printf("%s FBFC [%2.2x] FBFDw[%2.2x] FBFE [%2.2x] writenum [%4.4x]\n",machine().describe_context().c_str(),m_port0,data,m_fbfe,writenum);

		data &= m_fbfd_mask;

		// ignore repeated values and updates while bit 4 of port 0 is set
		// (counter goes crazy when bit 4 is set so must have a different purpose then)
		if ((data != m_fbfd) && !(m_port0 & 0x10))
		{
			m_old_old_fbfd=m_old_fbfd;
			m_old_fbfd=m_fbfd;
			m_fbfd=data;

			writenum++;

			//check_scroll_register();
			m_vram.set_scroll_register(m_fbfd);
```
Just tried it quickly and it seems to work fine.  I could make this optimisation now or another time.  I was planning to look into the character generator next and that also uses the same port.

Reply:
I've made another commit to remove redundant code and member variables.  Using the counter directly works really well and is of course what the real hardware did. Hopefully that should be everything unless you have any further comments.